File: cli/adapters/songs.py
"""Defines the Songs Adapter class"""

from typing import Any

from adapters.base import Adapter


class SongsAdapter(Adapter):
    """Defines an adapter pattern for the Songs model"""

    def make_table(self) -> list[list[Any]]:
        """
        Converts the songs data into a list of rows

        Returns
        -------
        table: list[list[Any]]
            a list of rows which are lists of song ranks, names, artists, popularity,
            album, release date, and IDs
        """
        return [
            [
                idx + 1,
                item["name"],
                ", ".join(item["artists"]),
                item["popularity"],
                item["album"],
                item["release_date"],
                item["id"],
            ]
            for idx, item in enumerate(self.data["items"])
        ]

File: rcli/src/models/songs.rs
use crate::models::content::Content;
use prettytable::Row;
use serde::Deserialize;

#[derive(Deserialize, Debug)]
pub struct Song {
    name: String,
    artists: Vec<String>,
    popularity: u8,
    album: String,
    release_date: String,
    id: String,
}

impl Content for Song {
    fn as_row(&self, idx: &usize) -> Row {
        row!(
            idx + 1,
            &self.name,
            &self.artists.join(", "),
            &self.popularity,
            &self.album,
            &self.release_date,
            &self.id
        )
    }
}
